File: matlab/test_run.m
function test_run()
% TEST_RUN 快速测试：加载20只股票，计算特征，生成信号，单股回测
%
%   用法:
%       test_run()

fprintf('快速测试 (MATLAB)\n\n');

% 添加路径
addpath(fileparts(mfilename('fullpath')));

cfg = config();
data_dir = cfg.DATA_DIR;

% 1. 加载数据
[stock_data, stock_codes] = load_all_stocks(data_dir, 20);
fprintf('\n加载了 %d 只股票\n', length(stock_data));

% 2. 测试特征计算
code = stock_data(1).code;
tbl = stock_data(1).data;
tbl = calculate_all_features(tbl);
fprintf('特征计算完成: %d 列\n', width(tbl));
colnames = tbl.Properties.VariableNames;
fprintf('列名: %s\n', strjoin(colnames, ', '));

% 检查非 NaN 值数量
check_cols = {'MA5', 'RSI14', 'DIF', 'box_pos_60', 'K'};
for c = 1:length(check_cols)
    col = check_cols{c};
    if any(strcmp(colnames, col))
        n_valid = sum(~isnan(tbl.(col)));
        fprintf('有效 %s: %d\n', col, n_valid);
    end
end

% 3. 测试信号生成
[strategies, strategy_keys, strategy_names] = get_all_strategies();
tbl = generate_all_signals(tbl, strategies, strategy_keys);
for s = 1:length(strategy_keys)
    sig_col = sprintf('signal_%s', strategy_keys{s});
    n_signals = sum(tbl.(sig_col));
    fprintf('%s 信号数: %d\n', strategy_names{s}, n_signals);
end

% 4. 测试单股回测
for s = 1:length(strategy_keys)
    key = strategy_keys{s};
    name = strategy_names{s};
    sig_col = sprintf('signal_%s', key);
    trades = backtest_single_stock(tbl, sig_col, 10);
    if ~isempty(fieldnames(trades))
        returns = [trades.net_return_pct];
        win_rate = mean(returns > 0) * 100;
        avg_ret = mean(returns);
        fprintf('\n%s (hold 10d): %d trades, win_rate(>0)=%.1f%%, avg_return=%.2f%%\n', ...
            name, length(trades), win_rate, avg_ret);
    else
        fprintf('\n%s (hold 10d): 无交易\n', name);
    end
end

fprintf('\n=== 测试通过! ===\n');

end

File: test_run.py
# -*- coding: utf-8 -*-
"""快速测试：加载20只股票，计算特征，生成信号"""

import sys
sys.path.insert(0, "stock_strategy")

from config import *
import config
config.MAX_STOCKS = 20
config.RESULT_DIR = "E:/数学实验项目/results_test"

from data_loader import load_all_stocks
stock_data = load_all_stocks("E:/数学实验项目/day", max_stocks=20)
print(f"\nLoaded {len(stock_data)} stocks\n")

# Test feature calculation on one stock
from features import calculate_all_features
code = list(stock_data.keys())[0]
df = calculate_all_features(stock_data[code])
print(f"Features calculated for {code}: {len(df.columns)} columns")
print(f"Columns: {list(df.columns)}")
print(f"Non-null MA5: {df['MA5'].notna().sum()}")
print(f"Non-null RSI14: {df['RSI14'].notna().sum()}")
print(f"Non-null DIF: {df['DIF'].notna().sum()}")
print(f"Non-null box_pos_60: {df['box_pos_60'].notna().sum()}")
print(f"Non-null K: {df['K'].notna().sum()}")

# Test signal generation
from strategy import generate_all_signals, get_all_strategies
strategies = get_all_strategies()
df = generate_all_signals(df, strategies)
for key, (_, name) in strategies.items():
    sig_col = f"signal_{key}"
    n_signals = df[sig_col].sum()
    print(f"Signals for {name}: {int(n_signals)}")

# Test backtest
from backtest import backtest_single_stock, calculate_metrics
for key, (_, name) in strategies.items():
    sig_col = f"signal_{key}"
    trades = backtest_single_stock(df, sig_col, hold_days=10)
    metrics = calculate_metrics(trades)
    print(f"\n{name} (hold 10d): {metrics['trade_count']} trades, "
          f"win_rate(>0)={metrics['win_rate_gt0']}%, "
          f"avg_return={metrics['avg_return']}%")

print("\n=== Test passed! ===")
